### PR TITLE
[WIP] CUDA implementation of direct DiracDeterminant inversion.

### PR DESCRIPTION
**Pull request by PDoakORNL**
## Proposed changes

Changes to DiracDeterminantBatched and related classes to integrate the batch direct inversion using cuBLAS.  Some refactoring of the relationship between the Determinant, Determinant Engine, and the Determinant Inverter have been done. These correspond to `DiracDeterminant`, `DelayedUpdate`, `DiracMatrix` in the legacy design.

This PR finishes the CUDA implementation but refactoring between DiracDeterminantBatched, and the CUDA and OMPTarget implementations in the two lower levels remains to be done. 

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- New feature
- Code style update (formatting, renaming)
- Refactoring (no functional changes, no api changes)
- Testing changes (e.g. new unit/integration/performance tests)
- Documentation changes

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Leconte

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes/No. This PR is up to date with current the current state of 'develop'
- Yes/No. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- Yes. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by prckent** [comment]
The obligatory missing file,  Fermion/MatrixUpdateOMPTarget.cpp

**Comment by PDoakORNL** [comment]
Working through some build issues with different flavor combinations.

File: src/tests/integration/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${QMCPACK_UNIT_TEST_DIR})
+
+set(SRC_DIR dualallocators)
+set(UTEST_DIR ${qmcpack_BINARY_DIR}/tests/integration)
+
+execute_process(COMMAND ${CMAKE_COMMAND} -E make_directory "${UTEST_DIR}")
+
+
+set(UTEST_EXE test_${SRC_DIR})
+set(UTEST_NAME deterministic-integration_${SRC_DIR})
```
Comment (ye-luo) [inline src/tests/integration/CMakeLists.txt:24]: Just leaving a note. I notice on develop.
```
786/816 Test   #31: deterministic-unit_test_hamiltonian_force ........................................................   Passed    1.42 sec
        Start   26: deterministic-unit_test_wavefunction_jastrow
787/816 Test   #26: deterministic-unit_test_wavefunction_jastrow .....................................................   Passed    1.34 sec
        Start    3: deterministic-integration_dualallocators
788/816 Test    #3: deterministic-integration_dualallocators .........................................................   Passed    0.58 sec
```
better to rename as `deterministic-unit_test_containers`

Comment (PDoakORNL) [reply src/tests/integration/CMakeLists.txt:24]: See #3318

**Comment by ye-luo** [comment]
Note: Please rename assignUpperRight after merging develop after #3322 got in

**Comment by PDoakORNL** [comment]
Oh well

**Pull request closed without merging**